src/main/java/com/google/sps/servlets/DashboardServlet.java

```diff
@@ -14,62 +14,125 @@
+
  /**
   * This route will return a JSON with the percentages of each interaction
   */
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
 
+    String requestStartTimestamp = RequestUtils.getParameter(request, "startTime");
+    String requestEndTimestamp = RequestUtils.getParameter(request, "endTime");
+
+    validateAndSetVariables(requestStartTimestamp, requestEndTimestamp);    
+    
     try {
-      HashMap<String, Double> interactionPercentages = calculatePercentages();
+      HashMap<String, Double> dataToSend = calculatePercentages();
       
-      String jsonToSend = convertToJson(interactionPercentages);
+      String jsonToSend = convertToJson(dataToSend);
 
+      response.addHeader("Access-Control-Allow-Origin", "*");
+      response.setStatus(HttpServletResponse.SC_OK);
       response.setContentType("application/json; charset=UTF-8");
       response.getWriter().println(jsonToSend);
     
     } catch (IllegalAccessException exception) {
+
       response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
     }
   }
 
  /**
-  * This function queries for and calculates the percentages of each interaction.
+  * This function queries for and calculates the percentages of each
+  * interaction.
+  *
+  * @param startTimestamp Timestamp for the beginning of the date range that is requested.
+  * @param endTimestamp Timestamp for the end of the date range that is requested.
   * @return A hashmap with the interactions as keys and the percentages as values.
   */
-  public HashMap<String, Double> calculatePercentages() throws IllegalAccessException {
+  private HashMap<String, Double> calculatePercentages() throws IllegalAccessException {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    PreparedQuery interactions = datastore.prepare(new Query(DBUtilities.INTERACTION_TABLE));
+
+    final Filter startTimestampFilter =  new FilterPredicate(Property.TIMESTAMP, FilterOperator.GREATER_THAN_OR_EQUAL, this.startTimestamp);
+    final Filter endTimestampFilter =  new FilterPredicate(Property.TIMESTAMP, FilterOperator.LESS_THAN_OR_EQUAL, this.endTimestamp);
+    
+    // combine the two filters with the 'and' operator 
+    final Query timedQuery = new Query(DBUtilities.INTERACTION_TABLE).setFilter(CompositeFilterOperator.and(startTimestampFilter, endTimestampFilter));
+    PreparedQuery interactions = datastore.prepare(timedQuery);
 
     Field[] allFields = Property.class.getDeclaredFields();
+    Property classInstance = new Property();    
   
-    HashMap<String, Double> interactionPercentages = new HashMap<>();
+    HashMap<String, Double> dataToSend = new HashMap<>();
 
-    for (Field field : allFields) {
-      String property = (String) field.get(new Property()); // gets the value of the field variable
+    // start iterating at 2 to skip timestamp and correlator properties
+    for (int i = 2; i < allFields.length; ++i){
```

Comment:
This seems bad.  If someone were to add a new parameter, or even just sort the values, this would stop working.  I think it would be better to keep another constant somewhere, thats an array of the fields to search for

Reply:
made a const array of the fields